app/src/main/java/org/mozilla/fenix/wallpapers/WallpaperFileManager.kt

```diff
@@ -47,16 +46,99 @@ class WallpaperFileManager(
+                if (landscapeFile.exists()) {
+                    Pair(portraitFile, landscapeFile)
+                } else {
+                    null
                 }
             }
+
+        val migratedWallpapers = legacyWallpapers?.map { pair ->
+            val name = pair.first.nameWithoutExtension
+            availableWallpapers.firstOrNull { it.name == name } ?: Wallpaper(
+                name,
+                Wallpaper.DefaultCollection,
+                null,
+                null
+            )
+        }
+
+        legacyWallpapers?.let {
+            migrateLegacyWallpapers(availableWallpapers, it, downloadThumbnail)
+        }
+        migratedWallpapers ?: listOf()
+    }
+
+    private fun migrateLegacyWallpapers(
+        availableWallpapers: List<Wallpaper>,
+        legacyWallpapers: List<Pair<File, File>>,
+        downloadThumbnail: (Wallpaper, (Throwable) -> Unit) -> Unit
+    ) = scope.launch {
+        legacyWallpapers.map { pair ->
+            val portraitFile = pair.first
+            val landscapeFile = pair.second
+            val name = portraitFile.nameWithoutExtension
+            val remoteWallpaper = availableWallpapers.firstOrNull { it.name == name }
+
+            // If there is a thumbnail available for the current wallpaper,
+            // attempt to download it
+            if (remoteWallpaper != null) {
+                downloadThumbnail(
+                    remoteWallpaper
+                ) {
+                    // If the thumbnail failed to download, use the portrait as thumbnail
+                    portraitFile.copyTo(
+                        File(
+                            storageRootDirectory,
+                            "wallpapers/${name.lowercase()}/thumbnail.png"
+                        )
+                    )
+                }
+            } else {
+                // If there is no thumbnail available for the wallpaper,
+                // use the portrait as thumbnail
+                portraitFile.copyTo(
+                    File(
+                        storageRootDirectory,
+                        "wallpapers/${name.lowercase()}/thumbnail.png"
+                    )
+                )
+            }
+            // Copy the portrait file
+            portraitFile.copyTo(
+                File(
+                    storageRootDirectory,
+                    "wallpapers/${name.lowercase()}/portrait.png"
+                )
+            )
+            // Copy the landscape file
+            landscapeFile.copyTo(
+                File(
+                    storageRootDirectory,
+                    "wallpapers/${name.lowercase()}/landscape.png"
+                )
+            )
         }
+        // Delete the remaining legacy files
+        File(storageRootDirectory, "wallpapers/portrait").deleteRecursively()
+        File(storageRootDirectory, "wallpapers/landscape").deleteRecursively()
```

Comment:
does there need to be any File exception handling for any of the copy/delete operations?